File: 1340-jump-game-v/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
private:
    int n;
    vector<int> memo;

    int solve(int i, const vector<int>& arr, int d) {
        // If already computed, return the cached result
        if (memo[i] != -1) {
            return memo[i];
        }

        int max_visited = 1;

        // Try jumping to the right
        for (int j = i + 1; j <= min(n - 1, i + d); ++j) {
            if (arr[j] >= arr[i]) {
                break; // Met a "wall", cannot jump here or further
            }
            max_visited = max(max_visited, 1 + solve(j, arr, d));
        }

        // Try jumping to the left
        for (int j = i - 1; j >= max(0, i - d); --j) {
            if (arr[j] >= arr[i]) {
                break; // Met a "wall", cannot jump here or further
            }
            max_visited = max(max_visited, 1 + solve(j, arr, d));
        }

        return memo[i] = max_visited;
    }

public:
    int maxJumps(vector<int>& arr, int d) {
        n = arr.size();
        memo.assign(n, -1);
        int ans = 0;

        // Find the maximum path starting from any index
        for (int i = 0; i < n; ++i) {
            ans = max(ans, solve(i, arr, d));
        }

        return ans;
    }
};

File: 1340-jump-game-v/solution.py
from typing import List
from functools import lru_cache

class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:
        n = len(arr)
        
        @lru_cache(None)
        def dfs(i: int) -> int:
            max_visited = 1
            
            # Jump to the right
            for j in range(i + 1, min(n, i + d + 1)):
                if arr[j] >= arr[i]:
                    break  # Blocked by an element >= arr[i]
                max_visited = max(max_visited, 1 + dfs(j))
                
            # Jump to the left
            for j in range(i - 1, max(-1, i - d - 1), -1):
                if arr[j] >= arr[i]:
                    break  # Blocked by an element >= arr[i]
                max_visited = max(max_visited, 1 + dfs(j))
                
            return max_visited
            
        # Try starting from each index and get the overall maximum
        return max(dfs(i) for i in range(n))